pkg/daemon/daemon.go

```diff
@@ -361,23 +330,39 @@ func (dn *Daemon) BindPodMounts() error {
 }
 
 func (dn *Daemon) runLoginMonitor(stopCh <-chan struct{}, exitCh chan<- error) {
-	sessionNewCh := dn.loginClient.Subscribe("SessionNew")
-	for {
-		select {
-		case <-stopCh:
-			return
-		case msg, ok := <-sessionNewCh:
-			if !ok {
-				glog.V(4).Info("Not adding the ssh accessed annotation because the logind SessionNew channel is closed")
+	cmd := exec.Command("journalctl", "-b", "-f", "-o", "cat", "MESSAGE_ID=" + sdMessageSessionStart)
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		exitCh <- err
+		return
+	}
+	if err := cmd.Start(); err != nil {
+		exitCh <- err
+		return
+	}
+	go func() {
+		<- stopCh
+		cmd.Process.Kill()
+	}()
+	go func() {
```

Comment:
This goroutine only runs once and exits, should keep reading from the stream right?

The code below should be inside a for loop, with a select receiving from stopCh and exiting in that case (and killing the process). The other case in the select should be a default case with the actual code to read from the stream.

Reply:
in code:

```diff
diff --git a/pkg/daemon/daemon.go b/pkg/daemon/daemon.go
index 199a480..058f7b8 100644
--- a/pkg/daemon/daemon.go
+++ b/pkg/daemon/daemon.go
@@ -14,6 +14,7 @@ import (
 	"strings"
 	"time"
 
+	imgref "github.com/containers/image/docker/reference"
 	ignv2 "github.com/coreos/ignition/config/v2_2"
 	ignv2_2types "github.com/coreos/ignition/config/v2_2/types"
 	"github.com/golang/glog"
@@ -25,7 +26,6 @@ import (
 	mcfgclientv1 "github.com/openshift/machine-config-operator/pkg/generated/clientset/versioned/typed/machineconfiguration.openshift.io/v1"
 	"github.com/pkg/errors"
 	"github.com/vincent-petithory/dataurl"
-	imgref "github.com/containers/image/docker/reference"
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/wait"
@@ -261,7 +261,7 @@ const (
 
 // detectThisBootSSHAccesses taints the node if we find a login before the daemon started up.
 func (dn *Daemon) detectThisBootSSHAccesses() error {
-	journalOutput, err := exec.Command("journalctl", "-b", "-o", "cat", "MESSAGE_ID=" + sdMessageSessionStart).CombinedOutput()
+	journalOutput, err := exec.Command("journalctl", "-b", "-o", "cat", "MESSAGE_ID="+sdMessageSessionStart).CombinedOutput()
 	if err != nil {
 		return err
 	}
@@ -330,7 +330,7 @@ func (dn *Daemon) BindPodMounts() error {
 }
 
 func (dn *Daemon) runLoginMonitor(stopCh <-chan struct{}, exitCh chan<- error) {
-	cmd := exec.Command("journalctl", "-b", "-f", "-o", "cat", "MESSAGE_ID=" + sdMessageSessionStart)
+	cmd := exec.Command("journalctl", "-b", "-f", "-o", "cat", "MESSAGE_ID="+sdMessageSessionStart)
 	stdout, err := cmd.StdoutPipe()
 	if err != nil {
 		exitCh <- err
@@ -340,29 +340,38 @@ func (dn *Daemon) runLoginMonitor(stopCh <-chan struct{}, exitCh chan<- error) {
 		exitCh <- err
 		return
 	}
+	worker := make(chan struct{})
 	go func() {
-		<- stopCh
-		cmd.Process.Kill()
-	}()
-	go func() {
-		buf := make([]byte, 1024)
-		l, err := stdout.Read(buf)
-		if err != nil {
-			if err == io.EOF {
+		for {
+			select {
+			case <-worker:
 				return
-			}
-			exitCh <- err
-			return
-		}
-		if l > 0 {
-			line := strings.Split(string(buf), "\n")[0]
-			glog.Infof("Detected a new login session: %s", line)
-			glog.Infof("Login access is discouraged! Applying annotation: %v", machineConfigDaemonSSHAccessAnnotationKey)
-			if err := dn.applySSHAccessedAnnotation(); err != nil {
-				exitCh <- err
+			default:
+				buf := make([]byte, 1024)
+				l, err := stdout.Read(buf)
+				if err != nil {
+					if err == io.EOF {
+						return
+					}
+					exitCh <- err
+					return
+				}
+				if l > 0 {
+					line := strings.Split(string(buf), "\n")[0]
+					glog.Infof("Detected a new login session: %s", line)
+					glog.Infof("Login access is discouraged! Applying annotation: %v", machineConfigDaemonSSHAccessAnnotationKey)
+					if err := dn.applySSHAccessedAnnotation(); err != nil {
+						exitCh <- err
+					}
+				}
 			}
 		}
 	}()
+	select {
+	case <-stopCh:
+		close(worker)
+		cmd.Process.Kill()
+	}
 }
 
 func (dn *Daemon) applySSHAccessedAnnotation() error {
@@ -384,6 +393,7 @@ func (dn *Daemon) runKubeletHealthzMonitor(stopCh <-chan struct{}, exitCh chan<-
 				failureCount++
 				if failureCount >= kubeletHealthzFailureThreshold {
 					exitCh <- fmt.Errorf("kubelet health failure threshold reached")
+					return
 				}
 			} else {
 				failureCount = 0 // reset failure count on success
```

Reply:
Oh I had completely missed that this function was already in a goroutine.

Reply:
also, sorry, the `return` after `exitCh <- err` in the read buffer error isn't needed, we still want to stay in the goroutine and keep retrying

Reply:
So I'm getting a conflict trying to apply your patch...can you verify you're rebased on master?  Or since you're rewriting it entirely, just paste the new function?

Reply:
Actually I have unpushed changes but anyways just toss the new function my way literally instead of a diff and I think that'd be easier.

Reply:
mmm I've used this PR to create the patch above :confused: